CONTRIBUTING.md

```diff
@@ -77,6 +77,80 @@ README for more details.
 - Once a pull request is good to go, the code reviewer is the one who merges
   it, then deletes the branch.
 
+## Controller design
+
+To maintain consistency throughout the project, we've come up with the following
+design rules for our Rails controllers:
+
+### RESTful controllers
+
+* Only use CRUD methods in controllers.
+
+* Prefer adding a new controller with one of the CRUD methods over creating a
+  custom method in an existing controller. For example, if your app allows a
+  user to update their email and their password on two different pages, instead of
+  using a single controller with methods called `update_email` and
+  `update_password`, create two controllers and name the methods `update`, i.e.
+  `EmailsController#update` and `PasswordsController#update`. See
+  http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/ for more about
+  this design pattern.
+
+### Lean controllers
+* Keep as much business logic out of controllers.
+
+* Use specialized classes to handle the operations
+  * These will be Form Objects for the most part, since
+  most of what the app does is process user input via a form submission, or
+  clicking a link in email that contains a token.
+
+* Form Object rules:
+  - Must have a single public method called `submit` that returns a [FormResponse] object.
+  - Must use ActiveModel validations to validate the user input.
+  - Must be placed in `app/forms`.
+
+* Examples of Form/Service Objects:
+  - [EmailConfirmationTokenValidator]
+  - [PasswordForm]
+
+* The basic outline of how a controller interacts with this class is:
+```ruby
+result = Class.new(user).submit(params) # this returns a FormResponse object
+# all the necessary analytics attributes are captured inside the Form Object
+analytics.track_event('Some Event Name', result.to_h)
+
+if result.success?
+  handle_success
+else
+  handle_failure
+end
+```
+
+* Only make one call to `analytics.track_event` at the beginning of the
+controller action, as opposed to one call when handling success and another
+when handling failure. The Form Object, when used properly, will return
+a FormResponse object that already tells us whether the action was
+successful or not.
+
+By applying this pattern consistently, creating new controllers and capturing
+analytics is straightforward, as is testing. It makes it easy for new
+developers who join the team to see what the expected design is and how
```

Comment:
Except that by spreading related logic across many files, it makes it more difficult for developers to get up to speed on how particular things get done. So there's still some tension between making simple things more difficult and hard things less difficult.